src/newgrf_townname.cpp

```diff
@@ -21,69 +21,45 @@
 
 #include "safeguards.h"
 
-static GRFTownName *_grf_townnames = nullptr;
+static std::vector<GRFTownName> _grf_townnames;
 static std::vector<StringID> _grf_townname_names;
 
 GRFTownName *GetGRFTownName(uint32 grfid)
 {
-	GRFTownName *t = _grf_townnames;
-	for (; t != nullptr; t = t->next) {
-		if (t->grfid == grfid) return t;
-	}
+	auto found = std::find_if(std::begin(_grf_townnames), std::end(_grf_townnames), [&grfid](const GRFTownName &t){ return t.grfid == grfid; });
+	if (found != std::end(_grf_townnames)) return &(*found);
```

Comment:
&* not particularly nice. Do we need to be returning pointers at all now? Can we do references/copy-ellision instead?

Reply:
Looks like it needs to return nullptr if it doesn't exist, so needs to be a pointer.

Reply:
The "proper" C++ ham would be std::optional I guess.